### PR TITLE
[secure-net] add timeouts

### PR DESCRIPTION
**Pull request by davidiw**
This improve reliability on services that do not gracefully cleanup sockets by enforcing a 30 second blocking period before trying again.

* LSR this is a no-op
* Consensus will treat all LSR errors as debug and wait for next op
(this is a todo), for now, it will result in a panic in many cases, but
that's better than requiring an admin to manually restart.
* LEC this is a no-op
* Consensus has a lot of different entry points, some of them are part
of bootup and should likely panic and others are loggable errors.

In the case of panics, these seem like optimizations to improve recovery
of the node. If we hit a panic due to a faulty service, this behavior is
better than manual restarting from an admin. At the same time the
outcome can be improved if we can gracefully handle timeouts / retries.

See https://github.com/libra/libra/issues/4963
I believe this should address the concerns, additional logging is up to the users of this framework to log errors: https://github.com/libra/libra/blob/master/consensus/safety-rules/src/remote_service.rs#L34

### CODE REVIEW

File: secure/net/src/lib.rs [hunk 1]
```diff
@@ -351,4 +363,57 @@ mod test {
...
+        client.write(&data1).unwrap();
+        let result1 = server.read().unwrap();
+        assert_eq!(data1, result1);
+
+        // Timedout
+        server.read().unwrap_err();
+
+        // New client, success, note the previous client connection is still active, the server is
+        // actively letting it go due to lack of activity.
+        let mut client2 = NetworkClient::new(server_addr);
```
Comment (gregnazario) [inline secure/net/src/lib.rs:386]: Let's hope it doesn't somehow become a flaky test with a 5 second timeout.

Comment (davidiw) [reply secure/net/src/lib.rs:386]: I originally had an ignore, but opted against it, if 5 seconds leads to flakiness, I'll feel sad! and we can just add ignore where appropriate.

**Comment by davidiw** [comment]
> is this the issue happening in pre-mainnet? I saw `Connection established to upstream` printed in validators but seems LSR didn't accept the connection for whatever reason.

That's the suspicion (see the tagged issue), I'd love to see a netstat on the safety rules side that shows that we had two connections in established mode.

The alternative of course is to have the health checks cause the service to restart but ...

**Comment by gregnazario** [comment]
> > is this the issue happening in pre-mainnet? I saw `Connection established to upstream` printed in validators but seems LSR didn't accept the connection for whatever reason.
> 
> That's the suspicion (see the tagged issue), I'd love to see a netstat on the safety rules side that shows that we had two connections in established mode.
> 
> The alternative of course is to have the health checks cause the service to restart but ...

Yeah, having health checks cause a service to restart can be pretty dangerous without some tuning to make sure it isn't restarting all the time.

File: secure/net/src/lib.rs [hunk 2]
```diff
@@ -351,4 +363,57 @@ mod test {
...
+        let result1 = server.read().unwrap();
+        assert_eq!(data1, result1);
+
+        // Timedout, it is hard to simulate a client receiving a write timeout
+        client.read().unwrap_err();
+
+        // Clean up old Server listener but keep the stream online. Start a new server, which will
+        // be the one the client now connects to.
+        server.listener = None;
+        let mut server2 = NetworkServer::new(server_addr);
```
Comment (zekun000) [inline secure/net/src/lib.rs:412]: does it work with the same client/server after timeout?

Comment (davidiw) [reply secure/net/src/lib.rs:412]: It should, will expand the test.

**Comment by davidiw** [comment]
Forgot to land for branch cut...

@bors-libra r+

**Pull request closed without merging**